### DISCUSSION TITLE
Prometheus DS: Support Google Authentication

### DISCUSSION THREAD

**Discussion by bwplotka** [grafana/grafana#69180] [Data Sources]
# Proposal

Hi! 👋🏽 

I would like to propose adding support for Prometheus data source to access APIs authenticated using Google Auth ([Oauth2 JWT](https://cloud.google.com/docs/authentication#oauth2) as well as [ADC](https://cloud.google.com/docs/authentication/provide-credentials-adc)). This can reuse exactly the same authenticated code and frontend as [the Google Cloud Monitoring data source](https://grafana.com/docs/grafana/latest/datasources/google-cloud-monitoring/google-authentication/).

<img width="1440" alt="image" src="https://github.com/grafana/grafana/assets/6950331/4677da1f-bfe2-445b-9ad2-f6aff30f3c09">

This feature could be enabled by environment variable or configuration field, similar to [AWS](https://docs.aws.amazon.com/prometheus/latest/userguide/AMP-onboard-query-standalone-grafana.html) and [Azure](https://github.com/grafana/grafana/issues/35857) available for enterprise and open source users.

**Let me know if such a feature would be accepted if contributed by community.** We would love to help. 🤗

Happy to discuss here or in [DMs](https://www.bwplotka.dev/) 👋🏽 

Disclaimer: I am Prometheus & Thanos maintainer and I work with Google Managed Prometheus Team.

### Motivation

The [Google Managed Prometheus](https://cloud.google.com/stackdriver/docs/managed-prometheus) brings compatible Prometheus APIs for Google Cloud, enabling usage of Prometheus data source. Since Grafana Prometheus data source does not support neither [Oauth2](https://github.com/grafana/grafana/discussions/40844) nor Google Auth, Grafana users require installing [proxy](https://github.com/GoogleCloudPlatform/prometheus-engine/tree/main/cmd/frontend) next to Grafana to use Prometheus data source.

Secondly, Prometheus data source already allows AWS and Azure auth behind flags, thus there is precedence in enabling popular cloud provider's support to Prometheus datasource.

### Alternatives

* Keep recommending proxy, which is a poor experience for both Grafana and Google users.
* Supporting [Oauth2](https://github.com/grafana/grafana/discussions/40844) in Prometheus Data Source. Definitely helps, but this proposal makes experience against Google even better. 
* Adding Prometheus support to [Google Cloud Monitoring](https://grafana.com/docs/grafana/latest/datasources/google-cloud-monitoring/). Some disadvantages:
  * This feels a bit wrong, as it will mean there are two ways of doing same thing.
  * More work for everybody, more maintenance given Prometheus changes over time.
  * We are likely to recommend Prometheus data source for accessing GCM data going forward.

**Comment by bfife-bsci** [comment]
I would be happy to use this feature once developed, instead of having the prometheus frontend sidecar running beside our Grafana as a roundabout means of getting access to Google Managed Prometheus.

IMO I don't think there is anything stopping you from submitting a PR as long as it meets the [contributing guidelines](https://github.com/grafana/grafana/blob/3150d804288b18d7c59bf3db66c32c32c29beb76/CONTRIBUTING.md).  Given your last point about recommending Prometheus over GCM, it seems this datasource would be very useful to that end :)

**Reply by bwplotka** [reply]
We can submit PR (anyone is welcome to do so indeed), but there is no guarantee anybody will review them. 

I want to ensure first this feature will be welcomed by the Grafana owners. It's not obvious - they might have a different plan for future of those data sources 🤗

**Reply by bfife-bsci** [reply]
Hi, just wondering if there's been any movement on this.  Barring built-in support, could this be released as a plugin data source?

**Comment by mhoran** [comment]
I have opened a feature request since discussions are going away: https://github.com/grafana/grafana/issues/79410.

**Comment by gelicia** [comment]
Thanks @mhoran for the feature request! Closing this discussion.
